session.go

```diff
@@ -450,10 +451,23 @@ func (s *session) onSessionError(readErr error, writeErr error) {
 		c.close(false, readErr, writeErr)
 	}
 	s.closeOnce.Do(func() {
```

Comment:
It's not quite clear to me why don't put this block into `s.Close()`. Once the underlying connection is closed in `s.Close()`, it's impossible to send anything.

Reply:
It also reminds me that the buffered data you mentioned in https://github.com/getlantern/lampshade/pull/11/files/d420860019b840845797614bb292a3c0f2500be3#r152271731 can't be sent out anyway. TCP has the same behavior except for the rarely used "half-close".

Reply:
> It's not quite clear to me why don't put this block into s.Close(). Once the underlying connection is closed in s.Close(), it's impossible to send anything.

That's a good point, and likely a bug in the initial implementation. In Go, the behavior of a net.Conn is that any data written to it via `Write()` before calling `Close()` will be sent. I think we want the same for lampshade.

Reply:
Oh, I think I see why this hasn't been a problem in practice. The `net.Conn` that we use is actually the stream. The only cases in which the session gets closed directly are either if there's a network error or we hit our idle timeout.